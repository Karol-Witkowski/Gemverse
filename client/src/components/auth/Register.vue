<template>
  <v-container pl-16>
    <v-card
      class="mt-12 mx-auto"
      max-width="500px"
    >
      <v-card-title>
        <span class="headline grey--text text--darken-2">Create account *disabled*</span>
      </v-card-title>
      <v-card-text>
        <v-container>
          <v-row>
            <v-col cols="12">
              <v-text-field
                label="E-mail address"
                required
                :rules="[rules.required, rules.email]"
                value=""
              />
            </v-col>
            <v-col cols="12">
              <v-text-field
                label="Username"
                required
                :rules="[rules.required]"
                value=""
              />
            </v-col>
            <v-col cols="12">
              <v-text-field
                hint="At least 6 character password"
                label="Password"
                type="password"
                required
                :rules="[rules.counter, rules.required]"
                value=""
              />
            </v-col>
          </v-row>
          All fields are required and case-sensitive
        </v-container>
      </v-card-text>
      <v-card-actions class="pb-4">
        <v-btn
          color="blue lighten-2"
          text
          to="/login"
          outlined
        >
          Back
        </v-btn>
        <v-spacer />
        <v-btn
          color="blue lighten-2"
          text
          outlined
          to="/roomlist"
        >
          Sign up
        </v-btn>
      </v-card-actions>
    </v-card>
    <div class="mt-6 text-center">
      <h4 class="grey--text mx-auto mb-2 text--darken-1">Already have an account? Sign in</h4>
      <v-btn
        color="blue lighten-1"
        text
        to="/login"
        outlined
      >
        Sign in
      </v-btn>
    </div>
  </v-container>
</template>

<script>

export default {
  name: 'Register',
  data() {
    return {
      rules: {
        required: (value) => !!value || 'Required.',
        counter: (value) => (value.length <= 40 && value.length >= 6) || 'At least 6 character password',
        email: (value) => {
          const pattern = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
          return pattern.test(value) || 'Invalid e-mail.';
        },
      },
    };
  },
};
</script>
