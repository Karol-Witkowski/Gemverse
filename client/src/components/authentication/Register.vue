<template>
  <v-container class="mt-12">
    <v-card
      class="mt-12 mx-auto"
      max-width="500px"
    >
      <v-card-title>
        <span class="headline grey--text text--darken-2">Create account</span>
      </v-card-title>
      <v-card-text>
        <v-container>
          <v-form
            ref="form"
            v-model="isFormValid"
          >
            <v-row class="mb-2">
              <v-col cols="12">
                <v-text-field
                  autofocus
                  :error-messages="usernameError"
                  :counter="15"
                  label="Username"
                  ref="username"
                  required
                  :rules="generalRules.concat(usernameRules)"
                  v-model.trim="username"
                  v-on:keyup="usernameError = ''"
                  v-on:keyup.enter="formValidation"
                />
              </v-col>
              <v-col cols="12">
                <v-text-field
                  :error-messages="emailError"
                  label="E-mail address"
                  ref="email"
                  required
                  :rules="generalRules.concat(emailRules)"
                  v-model.trim="email"
                  v-on:keyup="emailError = ''"
                  v-on:keyup.enter="formValidation"
                />
              </v-col>
              <v-col cols="12">
                <v-text-field
                  hint="Minimum length - 6 characters"
                  label="Password"
                  persistent-hint
                  required
                  :rules="generalRules.concat(passwordRules)"
                  type="password"
                  v-model.trim="password"
                  v-on:keyup.enter="formValidation"
                />
              </v-col>
            </v-row>
            <p>All fields are required and case-sensitive</p>
          </v-form>
        </v-container>
      </v-card-text>
      <v-card-actions class="pb-4">
        <v-btn
          color="primary"
          outlined
          text
          to="/"
        >
          back
        </v-btn>
        <v-spacer />
        <v-btn
          @click.prevent="formValidation"
          color="primary"
          :disabled="!isFormValid"
          outlined
          text
        >
          sign up
        </v-btn>
      </v-card-actions>
    </v-card>
    <div class="mt-6 text-center">
      <h4 class="grey--text mx-auto mb-2 text--darken-1">Already have an account? Sign in</h4>
      <v-btn
        color="blue lighten-1"
        outlined
        text
        to="/login"
      >
        sign in
      </v-btn>
    </div>
  </v-container>
</template>

<script>
import axios from 'axios';
import tokenSetter from '@/utils/authTokenSetter';

export default {
  name: 'Register',
  data() {
    return {
      email: '',
      password: '',
      username: '',
      isFormValid: false,
      errors: {},
      emailError: '',
      usernameError: '',
      generalRules: [
        (value) => !(/[ ]/.test(value)) || 'No blank spaces allowed',
        (value) => !!value || 'Required',
      ],
      emailRules: [
        (value) => (value.length >= 8 && value.length <= 128) || 'E-mail adress must be at least 8 characters long',
        (value) => {
          const pattern = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
          return pattern.test(value) || 'Invalid e-mail';
        },
      ],
      passwordRules: [
        (value) => (value.length >= 6 && value.length <= 128) || 'Password must be at least 6 characters long',
      ],
      usernameRules: [
        (value) => (value.length >= 3 && value.length <= 15) || 'Characters range: 3 - 15',
      ],
    };
  },

  methods: {
    createUser() {
      axios.post('http://localhost:3000/api/authentication/register', {
        email: this.email,
        password: this.password,
        username: this.username,
      })
        .then((response) => {
          localStorage.setItem('authenticationToken', response.data.token);
          tokenSetter(response.data.token);
          this.dispatchToken();
          this.$store.dispatch('saveUser', response.data.user);

          if (response.status === 201) {
            this.$router.push({
              name: 'RoomList',
            });
          }
        })
        .catch((error) => {
          console.log(error);
          if (error.response.data.errors.email) {
            this.emailError = error.response.data.errors.email.msg;
          }

          if (error.response.data.errors.username) {
            this.usernameError = error.response.data.errors.username.msg;
          }
        });
    },

    dispatchToken() {
      if (localStorage.getItem('authenticationToken')) {
        this.$store.dispatch('remitAuthState', true);
      } else {
        localStorage.clear();
        this.$store.dispatch('remitAuthState', false);
      }
    },

    formValidation() {
      if (this.isFormValid) {
        this.createUser();
      }
    },
  },
};
</script>
