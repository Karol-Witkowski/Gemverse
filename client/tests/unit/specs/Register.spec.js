import { createLocalVue, mount } from '@vue/test-utils';
import axios from 'axios';
import Vue from 'vue';
import Vuetify from 'vuetify';
import Register from '@/components/authentication/Register.vue';
import tokenSetter from '@/utils/authTokenSetter';

const mockStore = { dispatch: jest.fn() };
const localSetItem = jest.spyOn(window.localStorage.__proto__, 'setItem');
const localVue = createLocalVue();
const url = 'http://localhost:3000/api/authentication/register';
let vuetify;
let wrapper;
const error = {
  response: {
    data: {
      errors: {
        email: {
          msg: 'Email error',
        },
        username: {
          msg: 'Username error',
        },
      },
    },
  },
};
const response = {
  data: {
    auth:	true,
    user: {
      data: 'testData',
    },
    success: true,
    token: 'testToken',
  },
};

jest.mock('axios');
vuetify = new Vuetify();

describe('Implementation test for Register.vue - successful HTTP post', () => {
  beforeEach(() => {
    axios.post.mockResolvedValue(response);

    wrapper = mount(Register, {
      localVue,
      mocks: {
        $store: mockStore,
      },
      stubs: [
        'router-link',
        'router-view',
      ],
      vuetify,
      data() {
        return {
          email: '',
          emailError: '',
          isFormValid: false,
          password: '',
          username: '',
          usernameError: '',
        };
      },
    });
  });

  afterEach(() => {
    axios.post.mockReset();
    localSetItem.mockClear();
    wrapper.destroy();
  });

  it('Render correctly', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('Initializes with correct elements', () => {
    // Test buttons initial state
    expect(wrapper.findAll('.v-btn').length).toEqual(3);
    expect(wrapper.findAll('.v-btn').at(0).text()).toMatch('back');
    expect(wrapper.findAll('.v-btn').at(1).text()).toMatch('sign up');
    expect(wrapper.findAll('.v-btn').at(2).text()).toMatch('sign in');
    expect(wrapper.findAll('.v-btn').at(1).element.disabled).toBeTruthy();

    // Test inputs initial state
    expect(wrapper.findAll('.v-text-field').length).toEqual(3);
    expect(wrapper.findAll('input').at(0).text()).toEqual('');
    expect(wrapper.findAll('input').at(1).text()).toEqual('');
    expect(wrapper.findAll('input').at(2).text()).toEqual('');

    // Test validation initial state
    expect(wrapper.findAll('.v-messages').length).toEqual(3);
    expect(wrapper.findAll('.v-messages').at(0).text()).toEqual('');
    expect(wrapper.findAll('.v-messages').at(1).text()).toEqual('');
    expect(wrapper.findAll('.v-messages').at(2).text()).toEqual('Minimum length - 6 characters');
    expect(wrapper.vm.isformValid).toBeFalsy();
  });

  it('Enables error messages', async () => {
    await wrapper.setData({
      emailError: 'Email error',
      usernameError: 'Username error',
    });

    expect(wrapper.findAll('.v-messages').length).toEqual(3);
    expect(wrapper.findAll('.v-messages').at(0).text()).toEqual('Username error');
    expect(wrapper.findAll('.v-messages').at(1).text()).toEqual('Email error');
  });

  it('Enables sign up button when user data were set', async () => {
    await wrapper.setData({
      email: 'test@email.js',
      password: 'password',
      username: 'username',
    });

    // Check that the user data is properly set
    expect(wrapper.findAll('input').at(0).element.value).toEqual('username');
    expect(wrapper.findAll('input').at(1).element.value).toEqual('test@email.js');
    expect(wrapper.findAll('input').at(2).element.value).toEqual('password');
    expect(wrapper.vm.username).toBe('username');
    expect(wrapper.vm.email).toBe('test@email.js');
    expect(wrapper.vm.password).toBe('password');

    await Vue.nextTick();
    // Check if validation pass
    expect(wrapper.vm.isFormValid).toBeTruthy();

    // Check that the sign up button is active
    expect(wrapper.findAll('.v-btn').at(1).element.disabled).toBeFalsy();
  });

  it('Fail validation when only one field was entered', async () => {
    await wrapper.setData({
      username: 'username',
    });

    expect(wrapper.vm.isFormValid).toBeFalsy();
  });

  it('Fail validation when email address pattern is invalid', async () => {
    await wrapper.setData({
      email: 'wrongPatter',
    });

    expect(wrapper.vm.isFormValid).toBeFalsy();
  });

  it('Should sends post request with correct data on form submit', async () => {
    await wrapper.setData({
      email: 'test@email.js',
      password: 'password',
      username: 'username',
    });

    await Vue.nextTick();

    wrapper.vm.createUser();

    // Check if post was called
    expect(axios.post).toHaveBeenCalled();

    // Check if post was called once
    expect(axios.post).toHaveReturnedTimes(1);

    // Check if post was called with correct data
    expect(axios.post).toHaveBeenCalledWith(
      url,
      {
        email: 'test@email.js',
        password: 'password',
        username: 'username',
      },
    );
  });

  it('Should store user data and auth status after successful register', async () => {
    wrapper.vm.createUser();

    // Check if any action were dispatched
    expect(mockStore.dispatch).toHaveBeenCalled();

    // Check if two actions were dispatched
    expect(mockStore.dispatch).toHaveReturnedTimes(2);

    // Check auth state was dispatched with correct data
    expect(mockStore.dispatch).toHaveBeenNthCalledWith(
      1, 'remitAuthState', true,
    );

    // Check user data was dispatched with correct data
    expect(mockStore.dispatch).toHaveBeenNthCalledWith(
      2, 'saveUser', response.data.user,
    );
  });

  it('Should store token on successful register', async () => {
    wrapper.vm.createUser();

    await tokenSetter('token', true);

    expect(localSetItem).toHaveBeenCalled();
    expect(localSetItem).toHaveBeenNthCalledWith(
      1, 'authenticationToken', 'testToken',
    );
  });
});

describe('Implementation test for Register.vue - failed HTTP post', () => {
  beforeEach(() => {
    axios.post.mockRejectedValue(error);

    wrapper = mount(Register, {
      localVue,
      mocks: {
        $store: mockStore,
      },
      stubs: [
        'router-link',
        'router-view',
      ],
      vuetify,
      data() {
        return {
          email: '',
          emailError: '',
          isFormValid: false,
          password: '',
          usernameError: '',
          username: '',
        };
      },
    });
  });

  afterEach(() => {
    axios.post.mockReset();
    mockStore.dispatch.mockReset();
    wrapper.destroy();
  });

  it('Display error messages on HTTP post failure', async () => {
    await wrapper.vm.createUser();

    await Vue.nextTick();

    expect(wrapper.findAll('.v-messages').length).toEqual(3);
    expect(wrapper.findAll('.v-messages').at(0).text()).toEqual('Username error');
    expect(wrapper.findAll('.v-messages').at(1).text()).toEqual('Email error');
    expect(wrapper.findAll('.v-messages').at(2).text()).toEqual('Minimum length - 6 characters');
  });

  it('Does not dispatch data on failed HTTP post', () => {
    wrapper.vm.createUser();

    expect(mockStore.dispatch).not.toHaveBeenCalled();
  });

  it('Does not set token when a failed HTTP post occurs', () => {
    wrapper.vm.createUser();

    expect(localSetItem).not.toHaveBeenCalled();
  });
});

describe('Behavioral test for Register.vue - successful HTTP post', () => {
  beforeEach(() => {
    axios.post.mockResolvedValue(response);

    wrapper = mount(Register, {
      localVue,
      mocks: {
        $store: mockStore,
      },
      stubs: [
        'router-link',
        'router-view',
      ],
      vuetify,
      data() {
        return {
          email: '',
          isFormValid: false,
          password: '',
          username: '',
        };
      },
    });
  });

  afterEach(() => {
    axios.post.mockReset();
    wrapper.destroy();
  });

  it('Should not sends post request when inputs are empty', async () => {
    wrapper.findAll('.v-btn').at(1).trigger('click');

    expect(axios.post).not.toHaveBeenCalled();
  });

  it('Should sends post request with correct on form submit', async () => {
    await wrapper.findAll('input').at(0).setValue('username');
    await wrapper.findAll('input').at(1).setValue('test@email.js');
    await wrapper.findAll('input').at(2).setValue('password');

    await Vue.nextTick();

    wrapper.findAll('.v-btn').at(1).trigger('click');

    // Check if sing in button was clicked
    expect(axios.post).toHaveBeenCalled();

    // Check if post was called once
    expect(axios.post).toHaveReturnedTimes(1);

    // Check if post was called with correct data
    expect(axios.post).toHaveBeenCalledWith(
      url,
      {
        email: 'test@email.js',
        password: 'password',
        username: 'username',
      },
    );
  });

  it('Should store user data and auth status after successful register', () => {
    wrapper.findAll('.v-btn').at(1).trigger('click');

    // Check if any action was dispatched
    expect(mockStore.dispatch).toHaveBeenCalled();

    // Check if two actions was dispatched
    expect(mockStore.dispatch).toHaveReturnedTimes(2);

    // Check auth state was dispatched with correct data
    expect(mockStore.dispatch).toHaveBeenNthCalledWith(
      1, 'remitAuthState', true,
    );

    // Check user data was dispatched with correct data
    expect(mockStore.dispatch).toHaveBeenNthCalledWith(
      2, 'saveUser', response.data.user,
    );
  });

  it('Should store token on successful register', async () => {
    wrapper.findAll('.v-btn').at(1).trigger('click');

    await tokenSetter('token', true);

    expect(localSetItem).toHaveBeenCalled();
    expect(localSetItem).toHaveBeenNthCalledWith(
      1, 'authenticationToken', 'testToken',
    );
  });
});

describe('Behavioral test for Register.vue - failed HTTP post', () => {
  beforeEach(() => {
    axios.post.mockRejectedValue(error);

    wrapper = mount(Register, {
      localVue,
      mocks: {
        $store: mockStore,
      },
      stubs: [
        'router-link',
        'router-view',
      ],
      vuetify,
      data() {
        return {
          email: '',
          emailError: '',
          isFormValid: false,
          password: '',
          usernameError: '',
          username: '',
        };
      },
    });
  });

  afterEach(() => {
    localSetItem.mockClear();
    mockStore.dispatch.mockReset();
    wrapper.destroy();
  });

  it('Display error messages on HTTP post failure', async () => {
    await wrapper.findAll('input').at(0).setValue('username');
    await wrapper.findAll('input').at(1).setValue('test@email.js');
    await wrapper.findAll('input').at(2).setValue('password');

    await Vue.nextTick();

    await wrapper.findAll('.v-btn').at(1).trigger('click');

    await Vue.nextTick();

    expect(wrapper.findAll('.v-messages').length).toEqual(3);
    expect(wrapper.findAll('.v-messages').at(0).text()).toEqual('Username error');
    expect(wrapper.findAll('.v-messages').at(1).text()).toEqual('Email error');
  });

  it('Hide HTTP post error messages when user enter new data', async () => {
    await wrapper.setData({
      emailError: 'Password error',
      usernameError: 'Email error',
    });

    await wrapper.findAll('input').trigger('keyup');

    expect(wrapper.findAll('.v-messages').length).toEqual(3);
    expect(wrapper.findAll('.v-messages').at(0).text()).toEqual('');
    expect(wrapper.findAll('.v-messages').at(1).text()).toEqual('');
  });

  it('Does not dispatch data when a failed HTTP post occurs', () => {
    wrapper.findAll('.v-btn').at(1).trigger('click');

    expect(mockStore.dispatch).not.toHaveBeenCalled();
  });

  it('Does not set token when a failed HTTP post occurs', () => {
    wrapper.findAll('.v-btn').at(2).trigger('click');

    expect(localSetItem).not.toHaveBeenCalled();
  });
});
