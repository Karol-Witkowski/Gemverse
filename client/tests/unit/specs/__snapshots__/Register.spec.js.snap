// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Implementation test for Register.vue - successful HTTP post Render correctly 1`] = `
<div class="container mt-12">
  <div class="mt-12 mx-auto v-card v-sheet theme--light" style="max-width: 500px;">
    <div class="v-card__title"><span class="headline grey--text text--darken-2">Create account</span></div>
    <div class="v-card__text">
      <div class="container">
        <form novalidate="novalidate" class="v-form">
          <div class="row mb-2">
            <div class="col col-12">
              <div class="v-input theme--light v-text-field">
                <div class="v-input__control">
                  <div class="v-input__slot">
                    <div class="v-text-field__slot"><label for="input-4" class="v-label theme--light" style="left: 0px; position: absolute;">Username</label><input name="username" required="required" autofocus="autofocus" id="input-4" type="text"></div>
                  </div>
                  <div class="v-text-field__details">
                    <div class="v-messages theme--light">
                      <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                    </div>
                    <div class="v-counter theme--light">0 / 15</div>
                  </div>
                </div>
              </div>
            </div>
            <div class="col col-12">
              <div class="v-input theme--light v-text-field">
                <div class="v-input__control">
                  <div class="v-input__slot">
                    <div class="v-text-field__slot"><label for="input-7" class="v-label theme--light" style="left: 0px; position: absolute;">E-mail address</label><input name="email" required="required" id="input-7" type="text"></div>
                  </div>
                  <div class="v-text-field__details">
                    <div class="v-messages theme--light">
                      <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div class="col col-12">
              <div class="v-input theme--light v-text-field">
                <div class="v-input__control">
                  <div class="v-input__slot">
                    <div class="v-text-field__slot"><label for="input-10" class="v-label theme--light" style="left: 0px; position: absolute;">Password</label><input name="password" required="required" id="input-10" type="password"></div>
                  </div>
                  <div class="v-text-field__details">
                    <div class="v-messages theme--light">
                      <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper">
                        <div class="v-messages__message">Minimum length - 6 characters</div>
                      </transition-group-stub>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <p>All fields are required and case-sensitive</p>
        </form>
      </div>
    </div>
    <div class="v-card__actions pb-4">
      <router-link-stub class="v-btn v-btn--outlined v-btn--router v-btn--text theme--light v-size--default primary--text" name="back"><span class="v-btn__content">
        back
      </span></router-link-stub>
      <div class="spacer"></div> <button type="button" disabled="disabled" class="v-btn v-btn--disabled v-btn--outlined v-btn--text theme--light v-size--default" name="register"><span class="v-btn__content">
        sign up
      </span></button>
    </div>
  </div>
  <div class="mt-6 text-center">
    <h4 class="grey--text mx-auto mb-2 text--darken-1">Already have an account? Sign in</h4>
    <router-link-stub class="v-btn v-btn--outlined v-btn--router v-btn--text theme--light v-size--default blue--text text--lighten-1" name="login"><span class="v-btn__content">
      sign in
    </span></router-link-stub>
  </div>
</div>
`;
